File: app/src/main/java/org/isoron/uhabits/activities/habits/list/views/CheckmarkButtonView.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.activities.habits.list.views;

import android.content.*;
import android.content.res.*;
import android.graphics.*;
import android.support.annotation.*;
import android.text.*;
import android.util.*;
import android.view.*;

import org.isoron.uhabits.*;
import org.isoron.uhabits.activities.habits.list.controllers.*;
import org.isoron.uhabits.utils.*;

import static android.view.View.MeasureSpec.*;
import static org.isoron.uhabits.models.Checkmark.*;
import static org.isoron.uhabits.utils.AttributeSetUtils.*;

public class CheckmarkButtonView extends View
{
    private int color;

    private int value;

    private StyledResources styledRes;

    private TextPaint paint;

    private int lowContrastColor;

    private RectF rect;

    public CheckmarkButtonView(Context context)
    {
        super(context);
        init();
    }

    public CheckmarkButtonView(@Nullable Context ctx, @Nullable AttributeSet attrs)
    {
        super(ctx, attrs);
        init();

        if(ctx == null) throw new IllegalStateException();
        if(attrs == null) throw new IllegalStateException();

        int paletteColor = getIntAttribute(ctx, attrs, "color", 0);
        setColor(ColorUtils.getAndroidTestColor(paletteColor));

        int value = getIntAttribute(ctx, attrs, "value", 0);
        setValue(value);
    }

    public void setColor(int color)
    {
        this.color = color;
        postInvalidate();
    }

    public void setController(final CheckmarkButtonController controller)
    {
        setOnClickListener(v -> controller.onClick());
        setOnLongClickListener(v -> controller.onLongClick());
    }

    public void setValue(int value)
    {
        this.value = value;
        postInvalidate();
    }

    public void toggle()
    {
        value = (value == CHECKED_EXPLICITLY ? UNCHECKED : CHECKED_EXPLICITLY);
        performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
        postInvalidate();
    }

    @Override
    protected void onDraw(Canvas canvas)
    {
        super.onDraw(canvas);
        Resources resources = getResources();

        paint.setColor(value == CHECKED_EXPLICITLY ? color : lowContrastColor);
        int id = (value == UNCHECKED ? R.string.fa_times : R.string.fa_check);
        String label = resources.getString(id);
        float em = paint.measureText("m");

        rect.set(0, 0, getWidth(), getHeight());
        rect.offset(0, 0.4f * em);
        canvas.drawText(label, rect.centerX(), rect.centerY(), paint);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec)
    {
        Resources res = getResources();
        int height = res.getDimensionPixelSize(R.dimen.checkmarkHeight);
        int width = res.getDimensionPixelSize(R.dimen.checkmarkWidth);

        widthMeasureSpec = makeMeasureSpec(width, EXACTLY);
        heightMeasureSpec = makeMeasureSpec(height, EXACTLY);

        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    private void init()
    {
        setFocusable(false);

        Resources res = getResources();
        styledRes = new StyledResources(getContext());

        paint = new TextPaint();
        paint.setTypeface(InterfaceUtils.getFontAwesome(getContext()));
        paint.setAntiAlias(true);
        paint.setTextAlign(Paint.Align.CENTER);
        paint.setTextSize(res.getDimension(R.dimen.regularTextSize));

        rect = new RectF();
        color = ColorUtils.getAndroidTestColor(0);
        lowContrastColor = styledRes.getColor(R.attr.lowContrastTextColor);
    }
}

File: android/uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/CheckmarkButtonView.kt
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.activities.habits.list.views

import android.content.*
import android.graphics.*
import android.text.*
import android.view.*
import android.view.View.MeasureSpec.*
import com.google.auto.factory.*
import org.isoron.androidbase.activities.*
import org.isoron.androidbase.utils.*
import org.isoron.uhabits.*
import org.isoron.uhabits.core.models.Checkmark.*
import org.isoron.uhabits.core.preferences.*
import org.isoron.uhabits.utils.*

@AutoFactory
class CheckmarkButtonView(
        @Provided @ActivityContext context: Context,
        @Provided val preferences: Preferences
) : View(context),
    View.OnClickListener,
    View.OnLongClickListener {

    var color: Int = Color.BLACK
        set(value) {
            field = value
            invalidate()
        }

    var value: Int = 0
        set(value) {
            field = value
            invalidate()
        }

    var onToggle: () -> Unit = {}
    var onToggleWithOptions: () -> Unit = {}
    private var drawer = Drawer()

    init {
        isFocusable = false
        setOnClickListener(this)
        setOnLongClickListener(this)
    }

    fun performToggle() {
        onToggle()
        value = when (value) {
            UNCHECKED -> CHECKED_EXPLICITLY
            else -> UNCHECKED
        }
        performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
        invalidate()
    }

    fun performToggleWithOptions() {
        onToggleWithOptions()
        performHapticFeedback(HapticFeedbackConstants.KEYBOARD_PRESS)
    }

    override fun onClick(v: View) {
        if (preferences.isShortToggleEnabled) performToggle()
        else if (preferences.isAdvancedCheckmarksEnabled) performToggleWithOptions()
        else showMessage(R.string.long_press_to_toggle)
    }

    override fun onLongClick(v: View): Boolean {
        if (preferences.isShortToggleEnabled && preferences.isAdvancedCheckmarksEnabled) {
            performToggleWithOptions()
        }
        else performToggle()
        return true
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawer.draw(canvas)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val height = resources.getDimensionPixelSize(R.dimen.checkmarkHeight)
        val width = resources.getDimensionPixelSize(R.dimen.checkmarkWidth)
        super.onMeasure(width.toMeasureSpec(EXACTLY),
                        height.toMeasureSpec(EXACTLY))
    }

    private inner class Drawer {
        private val rect = RectF()
        private val lowContrastColor = sres.getColor(R.attr.lowContrastTextColor)

        private val paint = TextPaint().apply {
            typeface = getFontAwesome()
            isAntiAlias = true
            textAlign = Paint.Align.CENTER
            textSize = dim(R.dimen.smallTextSize)
        }

        fun draw(canvas: Canvas) {
            val lighterColor = ColorUtils.setAlpha(color, 0.5f)

            paint.color = when (value) {
                CHECKED_EXPLICITLY -> color
                UNCHECKED_EXPLICITLY_NECESSARY -> lowContrastColor
                UNCHECKED -> lowContrastColor
                else -> lighterColor
            }
            var unchecked_symbol = R.string.fa_times
            var implicitCheckedSymbol = R.string.fa_check
            if (preferences.isAdvancedCheckmarksEnabled) {
                unchecked_symbol = R.string.fa_question
                implicitCheckedSymbol = R.string.fa_question
            }
            val id = when (value) {
                SKIPPED_EXPLICITLY -> R.string.fa_skipped
                UNCHECKED -> unchecked_symbol
                UNCHECKED_EXPLICITLY_NECESSARY -> R.string.fa_times
                CHECKED_IMPLICITLY -> implicitCheckedSymbol
                else -> R.string.fa_check
            }
            val label = resources.getString(id)
            val em = paint.measureText("m")

            rect.set(0f, 0f, width.toFloat(), height.toFloat())
            rect.offset(0f, 0.4f * em)
            canvas.drawText(label, rect.centerX(), rect.centerY(), paint)
        }
    }
}

